File: src/background/scalingStatDistribution.js
import { ipcMain } from "electron";
import {
  SEARCH_SCALING_STAT_DISTRIBUTIONS,
  COUNT_SCALING_STAT_DISTRIBUTIONS,
  STORE_SCALING_STAT_DISTRIBUTION,
  FIND_SCALING_STAT_DISTRIBUTION,
  UPDATE_SCALING_STAT_DISTRIBUTION,
  DESTROY_SCALING_STAT_DISTRIBUTION,
  COPY_SCALING_STAT_DISTRIBUTION,
  GLOBAL_MESSAGE_BOX,
  GLOBAL_MESSAGE,
} from "../constants";

const { knex } = require("../libs/mysql");

ipcMain.on(SEARCH_SCALING_STAT_DISTRIBUTIONS, (event, payload) => {
  let queryBuilder = knex().select().from("foxy.dbc_scaling_stat_distribution");
  if (payload.ID) {
    queryBuilder = queryBuilder.where("ID", payload.ID);
  }
  if (payload.Stat) {
    queryBuilder = queryBuilder
      .where("StatID_1", payload.Stat)
      .orWhere("StatID_2", payload.Stat)
      .orWhere("StatID_3", payload.Stat)
      .orWhere("StatID_4", payload.Stat)
      .orWhere("StatID_5", payload.Stat)
      .orWhere("StatID_6", payload.Stat)
      .orWhere("StatID_7", payload.Stat)
      .orWhere("StatID_8", payload.Stat)
      .orWhere("StatID_9", payload.Stat)
      .orWhere("StatID_10", payload.Stat);
  }
  queryBuilder = queryBuilder
    .limit(50)
    .offset(payload.page != undefined ? (payload.page - 1) * 50 : 0);

  queryBuilder
    .then((rows) => {
      event.reply(SEARCH_SCALING_STAT_DISTRIBUTIONS, rows);
    })
    .catch((error) => {
      event.reply(`${SEARCH_SCALING_STAT_DISTRIBUTIONS}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(COUNT_SCALING_STAT_DISTRIBUTIONS, (event, payload) => {
  let queryBuilder = knex()
    .count("* as total")
    .from("foxy.dbc_scaling_stat_distribution");
  if (payload.ID) {
    queryBuilder = queryBuilder.where("ID", payload.ID);
  }
  if (payload.Stat) {
    queryBuilder = queryBuilder
      .where("StatID_1", payload.Stat)
      .orWhere("StatID_2", payload.Stat)
      .orWhere("StatID_3", payload.Stat)
      .orWhere("StatID_4", payload.Stat)
      .orWhere("StatID_5", payload.Stat)
      .orWhere("StatID_6", payload.Stat)
      .orWhere("StatID_7", payload.Stat)
      .orWhere("StatID_8", payload.Stat)
      .orWhere("StatID_9", payload.Stat)
      .orWhere("StatID_10", payload.Stat);
  }

  queryBuilder
    .then((rows) => {
      event.reply(COUNT_SCALING_STAT_DISTRIBUTIONS, rows[0].total);
    })
    .catch((error) => {
      event.reply(`${COUNT_SCALING_STAT_DISTRIBUTIONS}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(STORE_SCALING_STAT_DISTRIBUTION, (event, payload) => {
  let queryBuilder = knex()
    .insert(payload)
    .into("foxy.dbc_scaling_stat_distribution");

  queryBuilder
    .then((rows) => {
      event.reply(STORE_SCALING_STAT_DISTRIBUTION, rows);
    })
    .catch((error) => {
      event.reply(`${STORE_SCALING_STAT_DISTRIBUTION}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(FIND_SCALING_STAT_DISTRIBUTION, (event, payload) => {
  let queryBuilder = knex()
    .select()
    .from("foxy.dbc_scaling_stat_distribution")
    .where(payload);

  queryBuilder
    .then((rows) => {
      event.reply(
        FIND_SCALING_STAT_DISTRIBUTION,
        rows.length > 0 ? rows[0] : {}
      );
    })
    .catch((error) => {
      event.reply(`${FIND_SCALING_STAT_DISTRIBUTION}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(UPDATE_SCALING_STAT_DISTRIBUTION, (event, payload) => {
  let queryBuilder = knex()
    .table("foxy.dbc_scaling_stat_distribution")
    .where(payload.credential)
    .update(payload.scalingStatDistribution);

  queryBuilder
    .then((rows) => {
      event.reply(UPDATE_SCALING_STAT_DISTRIBUTION, rows);
    })
    .catch((error) => {
      event.reply(`${UPDATE_SCALING_STAT_DISTRIBUTION}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(DESTROY_SCALING_STAT_DISTRIBUTION, (event, payload) => {
  let queryBuilder = knex()
    .table("foxy.dbc_scaling_stat_distribution")
    .where(payload)
    .delete();

  queryBuilder
    .then((rows) => {
      event.reply(DESTROY_SCALING_STAT_DISTRIBUTION, rows);
    })
    .catch((error) => {
      event.reply(`${DESTROY_SCALING_STAT_DISTRIBUTION}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(COPY_SCALING_STAT_DISTRIBUTION, (event, payload) => {
  let id = undefined;
  let scalingStatDistribution = undefined;

  let idQueryBuilder = knex()
    .select("ID")
    .from("foxy.dbc_scaling_stat_distribution")
    .orderBy("ID", "desc");
  let findScalingStatDistributionQueryBuilder = knex()
    .select()
    .from("foxy.dbc_scaling_stat_distribution")
    .where(payload);
  Promise.all([
    idQueryBuilder.then((rows) => {
      id = rows[0].ID;
    }),
    findScalingStatDistributionQueryBuilder.then((rows) => {
      scalingStatDistribution = rows.length > 0 ? rows[0] : {};
    }),
  ])
    .then(() => {
      scalingStatDistribution.ID = id + 1;
      let queryBuilder = knex()
        .insert(scalingStatDistribution)
        .into("foxy.dbc_scaling_stat_distribution");
      queryBuilder
        .then((rows) => {
          event.reply(COPY_SCALING_STAT_DISTRIBUTION, rows);
          event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
        })
        .catch((error) => {
          event.reply(`${COPY_SCALING_STAT_DISTRIBUTION}_REJECT`, error);
          event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
        })
        .finally(() => {
          event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
        });
    })
    .catch((error) => {
      event.reply(`${COPY_SCALING_STAT_DISTRIBUTION}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, JSON.stringify(error));
    });
});

File: lib/model/scaling_stat_distribution.dart
class ScalingStatDistribution {
  final int id;
  final int statId0;
  final int statId1;
  final int statId2;
  final int statId3;
  final int statId4;
  final int statId5;
  final int statId6;
  final int statId7;
  final int statId8;
  final int statId9;
  final int bonus0;
  final int bonus1;
  final int bonus2;
  final int bonus3;
  final int bonus4;
  final int bonus5;
  final int bonus6;
  final int bonus7;
  final int bonus8;
  final int bonus9;
  final int maxlevel;

  const ScalingStatDistribution({
    this.id = 0,
    this.statId0 = 0,
    this.statId1 = 0,
    this.statId2 = 0,
    this.statId3 = 0,
    this.statId4 = 0,
    this.statId5 = 0,
    this.statId6 = 0,
    this.statId7 = 0,
    this.statId8 = 0,
    this.statId9 = 0,
    this.bonus0 = 0,
    this.bonus1 = 0,
    this.bonus2 = 0,
    this.bonus3 = 0,
    this.bonus4 = 0,
    this.bonus5 = 0,
    this.bonus6 = 0,
    this.bonus7 = 0,
    this.bonus8 = 0,
    this.bonus9 = 0,
    this.maxlevel = 0,
  });

  factory ScalingStatDistribution.fromJson(Map<String, dynamic> json) {
    return ScalingStatDistribution(
      id: json['ID'] ?? 0,
      statId0: json['StatID0'] ?? 0,
      statId1: json['StatID1'] ?? 0,
      statId2: json['StatID2'] ?? 0,
      statId3: json['StatID3'] ?? 0,
      statId4: json['StatID4'] ?? 0,
      statId5: json['StatID5'] ?? 0,
      statId6: json['StatID6'] ?? 0,
      statId7: json['StatID7'] ?? 0,
      statId8: json['StatID8'] ?? 0,
      statId9: json['StatID9'] ?? 0,
      bonus0: json['Bonus0'] ?? 0,
      bonus1: json['Bonus1'] ?? 0,
      bonus2: json['Bonus2'] ?? 0,
      bonus3: json['Bonus3'] ?? 0,
      bonus4: json['Bonus4'] ?? 0,
      bonus5: json['Bonus5'] ?? 0,
      bonus6: json['Bonus6'] ?? 0,
      bonus7: json['Bonus7'] ?? 0,
      bonus8: json['Bonus8'] ?? 0,
      bonus9: json['Bonus9'] ?? 0,
      maxlevel: json['Maxlevel'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'ID': id,
      'StatID0': statId0,
      'StatID1': statId1,
      'StatID2': statId2,
      'StatID3': statId3,
      'StatID4': statId4,
      'StatID5': statId5,
      'StatID6': statId6,
      'StatID7': statId7,
      'StatID8': statId8,
      'StatID9': statId9,
      'Bonus0': bonus0,
      'Bonus1': bonus1,
      'Bonus2': bonus2,
      'Bonus3': bonus3,
      'Bonus4': bonus4,
      'Bonus5': bonus5,
      'Bonus6': bonus6,
      'Bonus7': bonus7,
      'Bonus8': bonus8,
      'Bonus9': bonus9,
      'Maxlevel': maxlevel,
    };
  }
}
